src/Analyzers/Core/Analyzers/UseConditionalExpression/ForAssignment/UseConditionalExpressionForAssignmentHelpers.cs

```diff
@@ -49,9 +49,40 @@ internal static class UseConditionalExpressionForAssignmentHelpers
                 return false;
             }
 
+            // If both assignments are discards type check is required.
+            // Since discard can discard value of any type, after converting if statement to a conditional expression
+            // it can produce compiler error if types are not the same and there is no implicit conversion between them, e.g.:
+            // if (flag)
+            // {
+            //     _ = 5;
+            // }
+            // else
+            // {
+            //     _ = "";
+            // }
+            // This code can result in `_ = flag ? 5 : ""`, which immediately produces CS0173
+            if (trueAssignment?.Target is IDiscardOperation &&
+                falseAssignment?.Target is IDiscardOperation &&
+                !AreEqualOrHaveImplicitConversion(trueAssignment.Type, falseAssignment.Type, trueAssignment.SemanticModel!.Compilation))
+            {
+                return false;
+            }
+
             isRef = trueAssignment?.IsRef == true;
             return UseConditionalExpressionHelpers.CanConvert(
                 syntaxFacts, ifOperation, trueStatement, falseStatement);
+
+            static bool AreEqualOrHaveImplicitConversion(ITypeSymbol? firstType, ITypeSymbol? secondType, Compilation compilation)
+            {
+                if (Equals(firstType, secondType))
+                    return true;
+
+                if (firstType is null || secondType is null)
+                    return false;
+
+                return compilation.ClassifyCommonConversion(firstType, secondType).IsImplicit
+                    || compilation.ClassifyCommonConversion(secondType, firstType).IsImplicit;
```

Comment:
This logic looks incorrect. It should be an XOR rather than an OR.

Test case:

```csharp
void M(bool b)
{
    if (b)
    {
        _ = GetC();
    }
    else
    {
        _ = GetString();
    }

    // error CS0172: Type of conditional expression cannot be determined because 'C' and 'string' implicitly convert to one another
    _ = b ? GetC() : GetString();
}

C GetC() => new C();
string GetString() => "";
 
public class C
{
    public static implicit operator C(string c) => new C();
    public static implicit operator string(C c) => "";
}
```

Reply:
Very good catch, thanks!